File: lib/maifetch/render.ex
defmodule Maifetch.Render do
  @moduledoc false

  def colour(value), do: IO.ANSI.color(72, 184, 200) <> value <> IO.ANSI.reset()

  def wide_to_normal(value) do
    value
    |> String.to_charlist()
    |> Enum.map(fn code -> if code >= 0xFF01 and code <= 0xFF5E, do: code - 0xFEE0, else: code end)
    |> List.to_string()
  end

  def difficulty_string(value) do
    %{
      "basic" => "BASIC",
      "advanced" => "ADVANCED",
      "expert" => "EXPERT",
      "master" => "MASTER",
      "remaster" => "Re:MASTER"
    }
    |> Map.get(value, "UNKNOWN")
  end

  def rank_string(value) do
    %{
      "d" => "D",
      "c" => "C",
      "b" => "B",
      "bb" => "BB",
      "bbb" => "BBB",
      "a" => "A",
      "aa" => "AA",
      "aaa" => "AAA",
      "s" => "S",
      "sp" => "S+",
      "ss" => "SS",
      "ssp" => "SS+",
      "sss" => "SSS",
      "sssp" => "SSS+"
    }
    |> Map.get(value, "UNKNOWN")
  end

  def info_lines(profile, plays, score_count) do
    name = profile |> get_in(["name"]) |> to_string() |> wide_to_normal()

    base = [
      colour(name),
      String.duplicate("-", String.length(name)),
      "#{colour("ID")}: #{profile["id"]}",
      "#{colour("Rating")}: #{rating(profile["rating"])} / #{rating(profile["rating_highest"])}",
      "#{colour("Level")}: #{profile["level"]}",
      "#{colour("Total Credits")}: #{get_in(profile, ["play_stats", "total"])}",
      "#{colour("Recent Scores")}:"
    ]

    scores =
      plays
      |> Enum.take(score_count)
      |> Enum.flat_map(fn play ->
        fc_label = play["full_combo_label"] || ""
        song_name = get_in(play, ["song", "name", "en"]) || "Unknown"
        difficulty = play |> get_in(["difficulty_level", "value"]) |> difficulty_string()
        rank = play |> Map.get("rank") |> rank_string()

        [
          "  #{song_name}  #{difficulty}",
          "  #{play["score_formatted"]} #{play["achievement_formatted"]}% #{rank} #{fc_label}",
          ""
        ]
      end)

    base ++ scores
  end

  def output(profile, plays, logo_size, score_count) do
    lines = info_lines(profile, plays, score_count)

    if logo_size > 0 do
      profile
      |> get_in(["options", "icon", "png"])
      |> ascii_logo(logo_size)
      |> print_combined(lines, logo_size)
    else
      Enum.join(lines, "\n")
    end
  end

  defp rating(nil), do: "0.00"
  defp rating(value) when is_integer(value), do: :io_lib.format("~.2f", [value / 100]) |> IO.iodata_to_binary()
  defp rating(value) when is_float(value), do: :io_lib.format("~.2f", [value / 100]) |> IO.iodata_to_binary()
  defp rating(value), do: rating(String.to_integer(to_string(value)))

  defp ascii_logo(_url, size) do
    width = max(size * 2, 1)
    height = max(size, 1)
    label = "MaiTea"
    pad = max(div(width - String.length(label), 2), 0)
    middle = String.duplicate(" ", pad) <> label

    Enum.map(1..height, fn row ->
      if row == div(height, 2) + 1 do
        String.pad_trailing(middle, width)
      else
        String.duplicate(" ", width)
      end
    end)
  end

  defp print_combined(logo_lines, info_lines, logo_size) do
    max_len = max(length(logo_lines), length(info_lines))
    blank_logo = String.duplicate(" ", max(logo_size * 2, 0))

    0..(max_len - 1)
    |> Enum.map(fn idx ->
      logo = Enum.at(logo_lines, idx, blank_logo)
      info = Enum.at(info_lines, idx, "")
      "#{logo}  #{info}"
    end)
    |> Enum.join("\n")
  end
end

File: src/Maifetch/Render.fs
// LLM-assisted development: OpenAI GPT-5 Codex
namespace Maifetch

open System

module Render =
    let private esc = "\u001b["

    let rgb fgR fgG fgB bgR bgG bgB text =
        $"{esc}38;2;{fgR};{fgG};{fgB};48;2;{bgR};{bgG};{bgB}m{text}{esc}0m"

    let fg r g b text =
        $"{esc}38;2;{r};{g};{b}m{text}{esc}0m"

    let colour text = fg 72 184 200 text

    let wideToNormal (text: string) =
        text
        |> Seq.map (fun ch ->
            let code = int ch
            if code >= 0xFF01 && code <= 0xFF5E then char (code - 0xFEE0) else ch)
        |> Seq.toArray
        |> String

    let difficulty value =
        match value with
        | "easy" -> rgb 255 255 255 69 174 255 "Easy"
        | "basic" -> rgb 255 255 255 111 212 61 "Basic"
        | "advanced" -> rgb 255 255 255 248 183 9 "Advanced"
        | "expert" -> rgb 255 255 255 255 46 66 "Expert"
        | "master" -> rgb 255 255 255 171 140 233 "Master"
        | "remaster"
        | "re:master" -> rgb 255 255 255 207 114 237 "Re:Master"
        | "utage" -> rgb 255 255 255 255 68 1 "Utage"
        | other -> other

    let rank value =
        match value with
        | "SSS+" -> fg 255 200 54 "S" + fg 225 38 165 "S" + fg 73 64 233 "S" + fg 21 203 148 "+"
        | "SSS" -> fg 255 200 54 "S" + fg 232 39 148 "S" + fg 18 195 144 "S"
        | "SS+" -> rgb 248 200 75 143 71 33 "SS+"
        | "SS" -> rgb 248 200 75 143 71 33 "SS"
        | "S+" -> rgb 248 200 75 75 82 82 "S+"
        | "S" -> rgb 248 200 75 75 82 82 "S"
        | "AAA" -> fg 23 163 255 "AAA"
        | "AA" -> fg 23 163 255 "AA"
        | "A" -> fg 23 163 255 "A"
        | other -> other

    let infoLines profile plays scoreCount =
        let name = wideToNormal profile.Name
        let idLabel = colour "ID"
        let ratingLabel = colour "Rating"
        let levelLabel = colour "Level"
        let creditsLabel = colour "Total Credits"
        let scoresLabel = colour "Recent Scores"
        let scoreLines =
            plays
            |> List.truncate scoreCount
            |> List.collect (fun play ->
                let fc = play.FullComboLabel |> Option.defaultValue ""
                [ $"  {play.Song.Name.En}  {difficulty play.DifficultyLevel.Value}"
                  $"  {play.ScoreFormatted} {play.AchievementFormatted}%% {rank play.Rank} {fc}"
                  "" ])

        [ colour name
          String.replicate name.Length "-"
          $"{idLabel}: {profile.Id}"
          $"{ratingLabel}: {float profile.Rating / 100.0:F2} / {float profile.RatingHighest / 100.0:F2}"
          $"{levelLabel}: {profile.Level}"
          $"{creditsLabel}: {profile.PlayStats.Total}"
          $"{scoresLabel}:" ]
        @ scoreLines

    let printText profile plays scoreCount =
        infoLines profile plays scoreCount
        |> String.concat Environment.NewLine
        |> printfn "%s"
